interop: implement rpc-behavior for UnaryCall()

Aditya-Sood opened the pull request:
Implements the spec found at: https://github.com/grpc/grpc/blob/master/doc/xds-test-descriptions.md#server

Fixes #6288 

RELEASE NOTES: none

Aditya-Sood commented:
hi @easwars @zasweq
could you please review the changes?

if they look good I'll start on the additional `rpc-behavior` testcase

zasweq commented on interop/xds/server/server.go:
```diff
@@ -78,8 +83,47 @@ func (s *testServiceImpl) EmptyCall(ctx context.Context, _ *testpb.Empty) (*test
+		mdRPCBehavior = md.Get(xdsinterop.RPCBehaviorMetadataKey)
+	}
+
+	rpcBehaviorMetadata := make(map[string]bool)
```
I just realized this about the C++ implementation (I commented on theirs too) and this, this needs to be a slice. Using a map takes away the ordering of the comma separated values, which breaks this line from the gRFC: "The `rpc-behavior` header value can have multiple options separated by commas. In that case, the value should be split by commas and the options ***should be applied in the order specified***" see the Java implementation for deterministic list ordering: https://github.com/grpc/grpc-java/blob/42b4c61d5e869b1534808204bc71df84720f5087/interop-testing/src/main/java/io/grpc/testing/integration/XdsTestServer.java#L403.

zasweq commented on interop/xds/server/server.go:
```diff
@@ -78,8 +83,47 @@ func (s *testServiceImpl) EmptyCall(ctx context.Context, _ *testpb.Empty) (*test
+	} else if errCode, err := strconv.Atoi(headerValue[len(xdsinterop.ErrorCodePrefix):]); err != nil {
+		return status.New(codes.InvalidArgument, "Invalid format for rpc-behavior header: "+headerValue)
+	} else {
+		return status.New(codes.Code(errCode), "Rpc failed as per the rpc-behavior header value: "+headerValue)
```
Same here, please you formatting directives. Also, would prefer keeping the happy case in the if (i.e. switch if conditional to err == nil).

easwars replied:
You can find more info here: https://google.github.io/styleguide/go/decisions#indent-error-flow

zasweq commented on interop/xds/server/server.go:
```diff
@@ -78,8 +83,47 @@ func (s *testServiceImpl) EmptyCall(ctx context.Context, _ *testpb.Empty) (*test
+		splitVals := strings.Split(val, ",")
+
+		for _, behavior := range splitVals {
+			rpcBehaviorMetadata[strings.TrimSpace(behavior)] = true
```
When you switch to a deterministic slice, I don't think trimming white space is speced anywhere/present in other languages. Please drop this trimming.

Aditya-Sood replied:
I think I had added it because of the java implementation - https://github.com/grpc/grpc-java/blob/master/interop-testing/src/main/java/io/grpc/testing/integration/XdsTestServer.java#L64

and they are also omitting empty strings

do the java repo owners need to be informed to remove it as well?
cc @easwars

zasweq replied:
Commented on their implementation about whitespace: https://github.com/grpc/grpc-java/commit/78067cbace7a745fb9953aba153c13573a71024a#r126228081.

zasweq commented on interop/xds/server/server.go:
```diff
@@ -78,8 +83,47 @@ func (s *testServiceImpl) EmptyCall(ctx context.Context, _ *testpb.Empty) (*test
+		}
+	}
+
+	response.Hostname = s.hostname
```
Just set hostname on line 86 inline in the struct.

Aditya-Sood replied:
@zasweq I did it purposely because even in the cpp implementation as well they aren't assigning it when a status is found for the RPC behaviour: https://github.com/grpc/grpc/blob/be99673d069217def9036cc9ae0346da324ba373/test/cpp/interop/xds_interop_server.cc#L122

Aditya-Sood replied:
would you still like me to move it?

zasweq replied:
I don't think the c++ implementation is right. I'd say go ahead and move it.

zasweq replied:
Commented on the c++ implementation.

zasweq reviewed (changes requested):
go.sum doesn't need to be updated like that, and please fix vet. You can run VET_SKIP_PROTO=1 ./vet.sh from the home directory to run the vet script. Also will need to run interop tests against this before merging to make sure this doesn't break anything.

easwars commented on interop/xds/server/server.go:
```diff
@@ -78,8 +83,47 @@ func (s *testServiceImpl) EmptyCall(ctx context.Context, _ *testpb.Empty) (*test
+	response := &testpb.SimpleResponse{ServerId: s.serverID}
+
+	for headerValue := range getRPCBehaviorMetadata(ctx) {
+		if rpcStatus := getStatusForRPCBehaviorMetadata(headerValue); rpcStatus != nil {
```
For these two variables `headerValue` and `rpcStatus`, use shorter names. Something like `val` and `st` would be good enough. See https://google.github.io/styleguide/go/decisions#variable-names for reasoning behind when to prefer using short variable names.

easwars commented on interop/xds/server/server.go:
```diff
@@ -78,8 +83,47 @@ func (s *testServiceImpl) EmptyCall(ctx context.Context, _ *testpb.Empty) (*test
+	if !strings.HasPrefix(headerValue, xdsinterop.ErrorCodePrefix) {
+		return nil
+	} else if errCode, err := strconv.Atoi(headerValue[len(xdsinterop.ErrorCodePrefix):]); err != nil {
+		return status.New(codes.InvalidArgument, "Invalid format for rpc-behavior header: "+headerValue)
```
We have a `status.Newf()` instead.

Aditya-Sood commented:
hi @zasweq and @easwars 
thanks for the review, I've implemented the changes in the latest commit

fixed the `go.sum` issue (I needed to use the `-compat` flag when running `go mod tidy` locally)
also all tests run fine with the latest commit

* I had a query regarding the string constant "rpc-behavior" being declared inline - is the cost of resolving references in go significant enough to prefer avoiding a single definition (which is easier to maintain)?

* also there are two pending reviews above (concerning the logic for extracting rpc-behaviour metadata) in which I require your clarifications, kindly let me know

thank you for sharing the style guide as well, I will go through it

easwars commented on interop/xds/server/server.go:
```diff
@@ -78,8 +87,48 @@ func (s *testServiceImpl) EmptyCall(ctx context.Context, _ *testpb.Empty) (*test
+func getRPCBehaviorMetadata(ctx context.Context) []string {
+	md, ok := metadata.FromIncomingContext(ctx)
+	if !ok {
+		logger.Error("failed to receive metadata")
```
Log strings don't need to follow the rules for error strings about not having them capitalized. So, this could instead be: `logger.Error("Failed to retrieve metadata from incoming RPC context")`

easwars commented on interop/xds/server/server.go:
```diff
@@ -78,8 +87,48 @@ func (s *testServiceImpl) EmptyCall(ctx context.Context, _ *testpb.Empty) (*test
+
+	mdRPCBehavior := md.Get(rpcBehaviorMetadataKey)
+
+	rpcBehaviorMetadata := make([]string, 0)
```
What is the point of allocating a slice with a capacity of `0`. Instead you would do `var rpcBehaviorMetadata []string` or `rpcBehaviorMetadata := []string{}`. 

And since `append` works on `nil` slices, the first option is totally viable.

Aditya-Sood replied:
I forgot to refactor the initialisation when I changed it from map to slice, will correct it now

easwars commented on interop/xds/server/server.go:
```diff
@@ -78,8 +87,48 @@ func (s *testServiceImpl) EmptyCall(ctx context.Context, _ *testpb.Empty) (*test
+	}
+
+	mdRPCBehavior := md.Get(rpcBehaviorMetadataKey)
+
```
Nit: nix this newline please.

Aditya-Sood replied:
I generally tend to add a new line wherever the reader might benefit from a logical pause while going down the code

could you help me understand the style directive in play here?

easwars commented on interop/xds/server/server.go:
```diff
@@ -78,8 +87,48 @@ func (s *testServiceImpl) EmptyCall(ctx context.Context, _ *testpb.Empty) (*test
+	response := &testpb.SimpleResponse{ServerId: s.serverID}
+
+	for _, val := range getRPCBehaviorMetadata(ctx) {
+		if st := getStatusForRPCBehaviorMetadata(val); st != nil {
```
Are you not planning to implement the complete spec here? https://github.com/grpc/grpc/blob/master/doc/xds-test-descriptions.md#server

```
In addition, when handling requests, if the initial request metadata contains the rpc-behavior key, it should modify its handling of the request as follows:

If the value matches sleep-<int>, the server should wait the specified number of seconds before resuming behavior matching and RPC processing.
If the value matches keep-open, the server should never respond to the request and behavior matching ends.
If the value matches error-code-<int>, the server should respond with the specified status code and behavior matching ends.
If the value matches success-on-retry-attempt-<int>, and the value of the grpc-previous-rpc-attempts metadata field is equal to the specified number, the normal RPC processing should resume and behavior matching ends.
A value can have a prefix hostname=<string> followed by a space. In that case, the rest of the value should only be applied if the specified hostname matches the server's hostname.
The rpc-behavior header value can have multiple options separated by commas. In that case, the value should be split by commas and the options should be applied in the order specified. If a request has multiple rpc-behavior metadata values, each one should be processed that way in order.
```

zasweq replied:
Yeah, if you want please go ahead and implement the full spec. If not, then once this is merged I'll leave the issue open with the full bullets, and perhaps I can pick it up in the future. I guess I linked the C++ PR in the issue that only adds error code logic. Right now with this change, I don't think we can specify our interop tests use Go Server Side.

easwars commented on interop/xds/server/server.go:
```diff
@@ -78,8 +87,48 @@ func (s *testServiceImpl) EmptyCall(ctx context.Context, _ *testpb.Empty) (*test
+
+	errCode, err := strconv.Atoi(headerValue[len(errorCodePrefix):])
+	if err != nil {
+		return status.Newf(codes.InvalidArgument, "Invalid format for rpc-behavior header: %v", headerValue)
```
This and the next one, being error strings, should not be capitalized. See: https://google.github.io/styleguide/go/decisions#error-strings

zasweq commented on interop/xds/custom_lb_test.go:
```diff
@@ -64,14 +64,14 @@ func (s) TestCustomLB(t *testing.T) {
+			rpcBMD := md.Get("rpc-behavior")
 			if len(rpcBMD) != 1 {
-				errCh.Send(errors.New("only one value should be received for metadata key " + RPCBehaviorMetadataKey))
+				errCh.Send(errors.New("only one value should be received for metadata key " + "rpc-behavior"))
```
No need for concatenation here, it's one string. "...metadata key: rpc-behavior"

zasweq commented on interop/xds/custom_lb_test.go:
```diff
@@ -64,14 +64,14 @@ func (s) TestCustomLB(t *testing.T) {
 			wantVal := "error-code-0"
 			if rpcBMD[0] != wantVal {
-				errCh.Send(fmt.Errorf("metadata val for key \"%v\": got val %v, want val %v", RPCBehaviorMetadataKey, rpcBMD[0], wantVal))
+				errCh.Send(fmt.Errorf("metadata val for key \"%v\": got val %v, want val %v", "rpc-behavior", rpcBMD[0], wantVal))
```
inline the rpc behavior string here as well (i.e. put it in the string rather than as an argument to formatting directive).

easwars commented on interop/xds/server/server.go:
```diff
@@ -78,8 +93,76 @@ func (s *testServiceImpl) EmptyCall(ctx context.Context, _ *testpb.Empty) (*test
+			time.Sleep(time.Duration(sleep) * time.Second)
+
+		} else if strings.HasPrefix(headerVal, keepOpenVal) {
+			return nil, nil
```
Returning `nil, nil` will terminate the RPC. I don't think that is what we want. The spec says: `If the value matches keep-open, the server should never respond to the request and behavior matching ends.`

So, we need to hang here. Something like `<-ctx.Done()` would be more appropriate.

easwars commented on interop/xds/server/server.go:
```diff
@@ -78,8 +93,76 @@ func (s *testServiceImpl) EmptyCall(ctx context.Context, _ *testpb.Empty) (*test
+		if strings.HasPrefix(headerVal, sleepPfx) {
+			sleep, err := strconv.Atoi(headerVal[len(sleepPfx):])
+			if err != nil {
+				return response, status.Errorf(codes.InvalidArgument, "invalid format for rpc-behavior header: %v", headerVal)
```
Here and everywhere else, where you return an error, please consider returning `nil` for the response.

Aditya-Sood replied:
In the cpp implementation they are setting the server ID in the response for even error returns, so I had followed the same approach
will correct it now

easwars commented on interop/xds/server/server.go:
```diff
@@ -78,8 +93,76 @@ func (s *testServiceImpl) EmptyCall(ctx context.Context, _ *testpb.Empty) (*test
+			return nil, nil
+
+		} else if strings.HasPrefix(headerVal, errorCodePfx) {
+			errCd, err := strconv.Atoi(headerVal[len(errorCodePfx):])
```
s/errCd/code

easwars commented on interop/xds/server/server.go:
```diff
@@ -78,8 +93,76 @@ func (s *testServiceImpl) EmptyCall(ctx context.Context, _ *testpb.Empty) (*test
+			return response, status.Errorf(codes.Code(errCd), "rpc failed as per the rpc-behavior header value: %v", headerVal)
+
+		} else if strings.HasPrefix(headerVal, successOnRetryPfx) {
+			attempt, err := strconv.Atoi(headerVal[len(successOnRetryPfx):])
```
Maybe rename this as `wantRetry` and the other one as `curRetry`.

easwars commented on interop/xds/server/server.go:
```diff
@@ -78,8 +93,76 @@ func (s *testServiceImpl) EmptyCall(ctx context.Context, _ *testpb.Empty) (*test
+			if retry == attempt {
+				break
+			}
+
```
Please don't include newlines at the end of every else-if block. The same applies when you switch over to `switch` as I've asked for in another comment.

easwars commented on interop/xds/server/server.go:
```diff
@@ -78,8 +93,76 @@ func (s *testServiceImpl) EmptyCall(ctx context.Context, _ *testpb.Empty) (*test
+			}
+
+		} else if strings.HasPrefix(headerVal, hostnamePfx) {
+			headerHostname := strings.TrimSpace(headerVal[len(hostnamePfx):])
+
+			if headerHostname != s.hostname {
```
Eliminate the local variable here.

zasweq commented on interop/xds/server/server.go:
```diff
@@ -78,8 +93,76 @@ func (s *testServiceImpl) EmptyCall(ctx context.Context, _ *testpb.Empty) (*test
+	response := &testpb.SimpleResponse{ServerId: s.serverID, Hostname: s.hostname}
+
+	for _, headerVal := range getRPCBehaviorMetadata(ctx) {
+		if strings.HasPrefix(headerVal, sleepPfx) {
```
Optional nit: can you also quote the spec for each of the switch cases: https://github.com/grpc/grpc/blob/master/doc/xds-test-descriptions.md#server

Aditya-Sood replied:
thanks for pointing this out, I had completely forgotten about this subtle enhancement over the constant case statements in cpp/java

Aditya-Sood commented on interop/xds/custom_lb_test.go:
```diff
@@ -66,7 +66,7 @@ func (s) TestCustomLB(t *testing.T) {
 			rpcBMD := md.Get("rpc-behavior")
 			if len(rpcBMD) != 1 {
-				errCh.Send(errors.New("only one value received for metadata key rpc-behavior"))
+				errCh.Send(errors.New("only one value should be received for metadata key rpc-behavior"))
```
yes only to fix the error message, because it says "one value received" even though the `len(rpcBMD) != 1` check means that is not the case

not required to be a part of this PR but I think it's a typo that needs to be fixed?

easwars replied:
Sure, if we can have it as part of a separate PR, that would be nicer. That way, we can keep the discussion here in this PR to the implementation of `rpc-behavior` metadata.

Aditya-Sood replied:
I've created PR #6614 for this
thanks

Aditya-Sood commented:
hi @easwars, could you please review the changes?

easwars commented on interop/xds/server/server.go:
```diff
@@ -78,8 +93,89 @@ func (s *testServiceImpl) EmptyCall(ctx context.Context, _ *testpb.Empty) (*test
+behavior:
+	for _, headerVal := range getRPCBehaviorMetadata(ctx) {
+		switch {
+		//If the value matches "sleep-<int>", the server should wait the specified number of seconds before resuming behavior matching and RPC processing.
```
Please include a space between `//` and the first character of the comment.

Here and in every other comment in this PR, please wrap them at 80-cols max. See: https://google.github.io/styleguide/go/decisions#comment-line-length

easwars commented on interop/xds/server/server.go:
```diff
@@ -78,8 +93,89 @@ func (s *testServiceImpl) EmptyCall(ctx context.Context, _ *testpb.Empty) (*test
+		switch {
+		//If the value matches "sleep-<int>", the server should wait the specified number of seconds before resuming behavior matching and RPC processing.
+		case strings.HasPrefix(headerVal, sleepPfx):
+			{
```
Parenthesis are not required to group statements within a `case`.

easwars commented on interop/xds/server/server.go:
```diff
@@ -78,8 +93,89 @@ func (s *testServiceImpl) EmptyCall(ctx context.Context, _ *testpb.Empty) (*test
+			{
+				sleep, err := strconv.Atoi(headerVal[len(sleepPfx):])
+				if err != nil {
+					return nil, status.Errorf(codes.InvalidArgument, "invalid format for rpc-behavior header %v, must be '%v<int>' instead", headerVal, sleepPfx)
```
instead of `must be '%v<int>`, prefer using `must be 'sleep-<int>`. That way, the code is more readable.

Same comment applies to other error strings as well.

easwars commented on interop/xds/server/server.go:
```diff
@@ -78,8 +93,89 @@ func (s *testServiceImpl) EmptyCall(ctx context.Context, _ *testpb.Empty) (*test
 func (s *testServiceImpl) UnaryCall(ctx context.Context, in *testpb.SimpleRequest) (*testpb.SimpleResponse, error) {
+	response := &testpb.SimpleResponse{ServerId: s.serverID, Hostname: s.hostname}
+
+behavior:
```
Consider renaming this label to `forLoop`.

easwars commented on interop/xds/server/server.go:
```diff
@@ -78,8 +93,89 @@ func (s *testServiceImpl) EmptyCall(ctx context.Context, _ *testpb.Empty) (*test
+		// If the value matches "keep-open", the server should never respond to the request and behavior matching ends.
+		case strings.HasPrefix(headerVal, keepOpenVal):
+			{
+				<-ctx.Done()
```
There should probably be a `break forLoop` after this receive returns?

Aditya-Sood replied:
shall I add a `return nil, nil` instead?
assuming it is fed the request's context, if the Done channel returns it would mean that the request has been closed by the client, so the return values will be of no use anyways?

easwars replied:
Returning nils would still be the way to go, because the spec says: `the server should never respond to the request and behavior matching ends`. So, we don't want to go back to the loop once the context expires.

easwars commented on interop/xds/server/server.go:
```diff
@@ -78,8 +93,89 @@ func (s *testServiceImpl) EmptyCall(ctx context.Context, _ *testpb.Empty) (*test
+					break behavior
+				}
+			}
+		// A value can have a prefix "hostname=<string>" followed by a space. In that case, the rest of the value should only be applied
+		// if the specified hostname matches the server's hostname.
```
Based on my reading of https://github.com/grpc/grpc-java/blob/master/interop-testing/src/main/java/io/grpc/testing/integration/XdsTestServer.java#L320, it looks like we can't use a `switch` after all.

So, we need to handle cases like `hostname=foo sleep-5`, and we need to process the `sleep-5` only if our server's hostname is `foo`.

So, this loop basically has to be:
```golang
for <iterate through the list of rpc behavior metadata values> {
  // currVal stores current value
  if hostname prefix {
    if hostname does not match { continue }
    // update currVal and run through the rest of the matchers
  }
  if sleep prefix { ... }
  if call behavior success after number retries prefix { ... }
  if keep open prefix { ... }
  if error prefix { ... }
}
```

Apologies for the back and forth. But hopefully, we've figured out how this is supposed to work this time around :)

Aditya-Sood replied:
pushed a commit with the above change

easwars replied:
Fair enough.

zasweq commented:
Reopening since that other PR is part of this, but not all of this.

easwars commented on interop/xds/server/server.go:
```diff
@@ -53,6 +58,16 @@ var (
 
+const (
+	rpcBehaviorMdKey  = "rpc-behavior"
+	rpcAttemptsMdKey  = "grpc-previous-rpc-attempts"
```
Nit: Consider renaming this as `grpcPreviousRPCAttempts` or `grpcPreviousRPCAttemptsKey` or `grpcPreviousRPCAttemptsMDKey`, to more closely match the actual key name.

Also, it needs to be `MD` and not `Md` because of https://google.github.io/styleguide/go/decisions#initialisms.

Aditya-Sood commented:
pushed commit with code review changes

easwars commented on interop/xds/server/server.go:
```diff
@@ -78,8 +93,93 @@ func (s *testServiceImpl) EmptyCall(ctx context.Context, _ *testpb.Empty) (*test
+
+forLoop:
+	for _, headerVal := range getRPCBehaviorMetadata(ctx) {
+		// A value can have a prefix "hostname=<string>" followed by a space. In that case,
+		// the rest of the value should only be applied if the specified hostname matches the server's hostname.
```
It feels like some of these comments are still not wrapped at 80-cols. Could you please take care of it. Thanks.

Aditya-Sood replied:
done
also sorry about the inconvenience with the reviews - I thought they were straightforward enough to be considered 'resolved' after my changes, but I now realise that you would still have to verify them once anyways
thanks for your patience and help!

easwars commented on interop/xds/server/server.go:
```diff
@@ -53,6 +58,16 @@ var (
+	sleepPfx                     = "sleep-"
+	keepOpenVal                  = "keep-open"
+	errorCodePfx                 = "error-code-"
+	successOnRetryPfx            = "succeed-on-retry-attempt-"
```
Rename the const `succeedOnRetryPfx` as well?

Aditya-Sood replied:
thanks had missed this

easwars reviewed (approved):
LGTM modulo those minor nits.

Also, please don't mark comments as resolved. It should be the reviewer who should be marking them as resolved. Because if you mark them as resolved, I have to go and open each one of them to see if you have handled them satisfactorily. Instead, if they remained open, it would be easy for me to look at them one at a time and mark them resolved, as appropriate. Thanks for understanding.

easwars commented:
@zasweq : The changes look good to me mostly. I will leave it to you for the second review and eventual merging. Thanks.

zasweq commented on interop/xds/server/server.go:
```diff
@@ -78,8 +93,96 @@ func (s *testServiceImpl) EmptyCall(ctx context.Context, _ *testpb.Empty) (*test
+		splitVals := strings.Split(val, ",")
+		rpcBehaviorMetadata = append(rpcBehaviorMetadata, splitVals...)
+	}
+
```
Nit: nix newline please

zasweq commented on interop/xds/server/server.go:
```diff
@@ -78,8 +93,96 @@ func (s *testServiceImpl) EmptyCall(ctx context.Context, _ *testpb.Empty) (*test
+		logger.Error("Failed to retrieve metadata from incoming RPC context")
+		return nil
+	}
+
```
Same here please, nix newline please

zasweq commented on interop/xds/server/server.go:
```diff
@@ -78,8 +93,96 @@ func (s *testServiceImpl) EmptyCall(ctx context.Context, _ *testpb.Empty) (*test
+		if strings.HasPrefix(headerVal, hostnamePfx) {
+			splitVal := strings.Split(headerVal, " ")
+			if len(splitVal) <= 1 {
+				return nil, status.Errorf(codes.InvalidArgument, "invalid format for rpc-behavior header %v, must be 'hostname=<string> <header>=<value>' instead", headerVal)
```
Here and other errors: can you please log the header val received so it can be compared to the correct format.

Aditya-Sood replied:
do you mean print a separate log along with returning the error?
because the error strings include `headerVal` as of now

zasweq commented on interop/xds/server/server.go:
```diff
@@ -78,8 +93,96 @@ func (s *testServiceImpl) EmptyCall(ctx context.Context, _ *testpb.Empty) (*test
+	mdRPCBehavior := getMetadataValues(ctx, rpcBehaviorMDKey)
+	var rpcBehaviorMetadata []string
+	for _, val := range mdRPCBehavior {
+		splitVals := strings.Split(val, ",")
```
What should we do with respect to whitespaces? Just leave them in the split values?

Aditya-Sood replied:
since they aren't in the spec maybe we shouldn't add it, as discussed earlier? https://github.com/grpc/grpc-go/pull/6575#discussion_r1304958572

although they don't seem to have seen your comment on the java implementation yet

zasweq replied:
Hello, we discussed this in our gRPC-Go team meeting today. The behavior we agreed upon is actually the trimming of the whitespace, as Java does it. Sorry for the hassle, I know I had you switch it earlier in this PR. Can you please switch it back then I'll go ahead and merge this. After merging, I'll file an issue in the gRPC repo to switch interop tests over to use gRPC-Go as the server. Thanks!

Aditya-Sood replied:
hi @zasweq, I've pushed the changes for removing whitespaces and ignoring empty strings

could you also clarify on your previous review about "log the header val"? - https://github.com/grpc/grpc-go/pull/6575#discussion_r1327836692

thanks

zasweq reviewed (approved):
LGTM modulo nits.

zasweq reviewed (changes requested):
One final thing

Aditya-Sood commented:
hi @zasweq, could you please clarify on the above query?

zasweq commented:
Vet proto seems to be failing, this looks good to merge though. Can you please rebase onto master and once it passes I'll merge?

Aditya-Sood commented:
hi @zasweq, vet proto is passing after the rebase
